[Excel](Custom Functions) Updates to CFs for Shared Runtime (#1835)
OfficeDev/office-js-docs-pr: docs/excel/custom-functions-authentication.md

---
ms.date: 05/13/2020
description: 'Authenticate users using UI-less custom functions in Excel.'
title: Authentication for UI-less custom functions
localization_priority: Normal
---

# Authentication for UI-less custom functions

> 

Reviewer: I feel like we should have some note above all the UI-less custom functions articles pointing people to the shared runtime. Or do we think people should still opt to use this pattern with equal frequency?

In some scenarios your UI-less custom function will need to authenticate the user in order to access protected resources. While custom functions don't require a specific method of authentication, be aware that UI-less custom functions run in a JavaScript-only runtime. Because of this, you'll need to pass data back and forth between the JavaScript-only runtime and the typical browser engine runtime used by most add-ins using the `OfficeRuntime.storage` object and the Dialog API.

[!include[Excel custom functions note](../includes/excel-custom-functions-note.md)]

## OfficeRuntime.storage object

> The JavaScript-only runtime used by UI-less custom functions doesn't have a `localStorage` object available on the global window, where you might typically store data. Instead, you should share data between UI-less custom functions and task panes by using [OfficeRuntime.storage](/javascript/api/office-runtime/officeruntime.storage) to set and get data.

Reviewer: ```suggestion
The JavaScript-only runtime used by UI-less custom functions doesn't have a `localStorage` object available on the global window, where you typically store data. Instead, you should share data between UI-less custom functions and task panes by using [OfficeRuntime.storage](/javascript/api/office-runtime/officeruntime.storage) to set and get data.
```

### Suggested usage

When you need to authenticate from a UI-less custom function, check `storage` to see if the access token was already acquired. If not, use the dialog API to authenticate the user, retrieve the access token, and then store the token in `storage` for future use.

## Dialog API

If a token doesn't exist, you should use the Dialog API to ask the user to sign in. After a user enters their credentials, the resulting access token can be stored in `storage`.

> [!NOTE]

> > The JavaScript-only runtime uses a Dialog object that is slightly different from the Dialog object in the browser engine runtime used by task panes. They're both referred to as the "Dialog API", but use `OfficeRuntime.Dialog` to authenticate users in the JavaScript-only runtime.

Reviewer: Consider this to make the takeaway more clear.
```suggestion
> The JavaScript-only runtime uses a Dialog object that is slightly different from the Dialog object in the browser engine runtime that is used by task panes (please note that both are referred to as the "Dialog API"). Use the `OfficeRuntime.Dialog` to authenticate users in the JavaScript-only runtime.
```

## General guidance

Office Add-ins are web-based and you can use any web authentication technique. There is no particular pattern or method you must follow to implement your own authentication with UI-less custom functions. You may wish to consult the documentation about various authentication patterns, starting with [this article about authorizing via external services](../develop/auth-external-add-ins.md).  

Avoid using the following locations to store data when developing custom functions:  

- `localStorage`: UI-less custom functions do not have access to the global `window` object and therefore have no access to data stored in `localStorage`.
- `Office.context.document.settings`:  This location is not secure and information can be extracted by anyone using the add-in.

## Dialog box API example

> In the following code sample, the function `getTokenViaDialog` uses the `Dialog` API's `displayWebDialogOptions` function to display a dialog box. This sample is provided only to show the capabilities of the `Dialog` object.

Reviewer: I live the callout in the code about this not showing the full authentication story. We should include that here.
```suggestion
In the following code sample, the function `getTokenViaDialog` uses the `Dialog` API's `displayWebDialogOptions` function to display a dialog box. This sample is provided to show the capabilities of the `Dialog` object, not demonstrate how to authenticate.
```

```JavaScript
/**
 * Function retrieves a cached token or opens a dialog box if there is no saved token. Note that this is not a sufficient example of authentication but is intended to show the capabilities of the Dialog object.
 * @param {string} url URL for a stored token.
 */
function getTokenViaDialog(url) {
  return new Promise (function (resolve, reject) {

>     if (_dialogOpen) {

Reviewer: What's up with the `_` before the variable names? Is that a standard in these APIs?

Author: It's a JS thing that can mean two things: it's private or an unused variable

>       // Can only have one dialog box open at once, wait for previous dialog box's token

Reviewer: Consider splitting this comment into two sentences.
```suggestion
      // Can only have one dialog box open at once. Wait for previous dialog box's token.
```
